### PR TITLE
Restrict Notifications to secure contexts.

### LINKED ISSUE #93: Restrict the Notifications API to secure contexts
Chromium is moving to [restrict access to the notifications API over insecure contexts](https://groups.google.com/a/chromium.org/forum/#!topic/Blink-dev/IVgkxkRNtMo). There will be a deprecation warning in M58, scheduled for April 2017, with full removal in M60, scheduled for August 2017. The proposed behaviour is that calls to Notification.requestPermission() and the Notification constructor from an insecure context will act as though the user has denied permission (similar to the behaviour of the geolocation API from an insecure context).

Motivations for the restriction are as follows:

 * The notification permission acceptance rate for HTTPS sites is 3x greater than for HTTP sites in Chrome
 * Over 90% of daily notification permission requests in Chrome are already made from a secure context
 * Notifications are overwhelming produced in tandem with the Push API which is [already restricted to secure contexts](https://w3c.github.io/push-api/#security-and-privacy-considerations). This restriction is only relevant for sites displaying local notifications while they are open in a browser tab
 * Aligning with the requirements of the Push API reduces developer friction and improves platform predictability
 * Chromium already requires local notifications on Android to be requested over a secure context
 * Notifications allow access to privileged system UI, making them a powerful feature

In light of this, we'd like to propose an update to the spec to explicitly require that user agents only make the Notifications API available over secure contexts, and to codify the behaviour of the API when requested from an insecure context. We also welcome suggestions on our proposed behaviour and whether it could be improved.

### PR DESCRIPTION
**Pull request by miketaylr**
Continuing the mess I made in #175.

Fixes #93

***
<a href="https://whatpr.org/notifications/176.html" title="Last updated on Nov 22, 2021, 7:57 PM UTC (4eff822)">Preview</a> | <a href="https://whatpr.org/notifications/176/df1b6db...4eff822.html" title="Last updated on Nov 22, 2021, 7:57 PM UTC (4eff822)">Diff</a>

### CODE REVIEW

**Comment by miketaylr** [comment]
(this will be even simpler with https://github.com/w3c/permissions/commit/a2464fbc53b4a114e15a923b05bd589444038856)

**Comment by miketaylr** [comment]
PTAL @annevk @marcoscaceres

File: notifications.bs [hunk 1]
```diff
@@ -291,37 +291,21 @@ removed from the <a>list of notifications</a>.
...
+ <li><p>If <var>permissionState</var> is "<code>prompt</code>", then return "<code>default</code>".
 
-<p>The Notifications API is a <a>powerful feature</a> which is identified by the string
-"notifications" and has its <a for="powerful feature">allowed in non-secure contexts</a> flag set.
-[[!Permissions]]
+ <li><p>Return <var>permissionState</var>.
+</ol>
 
 <p class="note">This specification's "<code>default</code>" <a>permission</a>
 state maps to the [=permission/prompt=] permission state in the Permissions API.
```
Comment (annevk) [inline notifications.bs:311]: Let's remove this as well, especially no we longer define permission.
```suggestion
```

**Review by annevk — APPROVED** [review]
This looks great. Just a couple of nits.

**Comment by annevk** [comment]
I tried to think through this again. How come `Notification.permission` does not end up prompting, but `Notification.requestPermission()` does? Wouldn't that require passing some kind of flag when obtaining the permission? The other places shouldn't end up prompting either, by the way.

File: notifications.bs [hunk 2]
```diff
@@ -291,40 +291,21 @@ removed from the <a>list of notifications</a>.
...
-  <dt>"<code>denied</code>"
-  <dd><p>This means the user does not want
-  <a>notifications</a>.
+<h3 id="permissions-integration">Permissions integration</h3>
 
-  <dt>"<code>granted</code>"
-  <dd><p>This means <a>notifications</a> can be
-  displayed.
-</dl>
+<p>The Notifications API is a <a>powerful feature</a> which is identified by the
```
Comment (annevk) [inline notifications.bs:296]:
```suggestion
<p>The Notifications API is a <a>powerful feature</a> which is identified by the
```
My bad...

Comment (miketaylr) [reply notifications.bs]: Heh, all good. Just pushed a fix.

**Comment by miketaylr** [comment]
> I tried to think through this again. How come `Notification.permission` does not end up prompting, but `Notification.requestPermission()` does?

Hmm, you're right - this isn't quite what we want. I think we should call https://w3c.github.io/permissions/#dfn-request-permission-to-use directly inside `requestPermission` instead of `get the notifications permission state`. The other places should be fine then.

File: notifications.bs [hunk 3]
```diff
@@ -760,27 +740,14 @@ method steps are:
...
-   <li><p>Run the <a for="powerful feature">permission query algorithm</a> with
-   <var>permissionDescriptor</var> and <var>permissionStatus</var>.
-
-   <li><p>Let <var>permissionState</var> be <var>permissionStatus</var>'s
-   {{PermissionStatus/state}}.
-
-   <li><p>If <var>permissionState</var> is {{PermissionState/"prompt"}}, then set
-   <var>permissionState</var> to "<code>default</code>".
+   <li><p>Let <var>permissionState</var> be the result of
+   <a>requesting permission to use</a> {{PermissionName/notifications}}.
```
[the remark this answers is not in the document]
Comment (miketaylr) [reply notifications.bs:750]: Yep, it only returns the strings "granted" or "denied":
> This algorithm returns either "granted" or "denied".

**Review by annevk — APPROVED** [review]

**Comment by annevk** [comment]
Thanks a lot @miketaylr for getting this cleaned up!

**Comment by miketaylr** [comment]
> Thanks a lot @miketaylr for getting this cleaned up!

Thanks for patiently guiding me through it. :)

**Pull request merged**